--- a/linux/kernel/bpf/syscall.c
+++ b/linux/kernel/bpf/syscall.c
@@ -1040,7 +1040,9 @@ static int map_lookup_elem(union bpf_att
 	if (attr->flags & ~BPF_F_LOCK)
 		return -EINVAL;
 
+	printk("User FD %d\n", ufd);
 	f = fdget(ufd);
+	printk("Kernel FD %p %u\n", f.file, f.flags);
 	map = __bpf_map_get(f);
 	if (IS_ERR(map))
 		return PTR_ERR(map);
@@ -1087,6 +1089,111 @@ err_put:
 	return err;
 }
 
+int accel_map_test_fn(int ufd)
+{
+	struct bpf_map *map;
+	struct fd f;
+
+	printk("User FD %d\n", ufd);
+	f = fdget(ufd);
+	printk("Kernel FD %p %u\n", f.file, f.flags);
+	map = __bpf_map_get(f);
+	if (IS_ERR(map))
+	{
+		printk("IS ERROR == TRUE\n");
+		return -1;
+	}
+
+	printk("Found map !!!!!\n");
+	return 0;
+}
+
+EXPORT_SYMBOL(accel_map_test_fn);
+
+struct fd accel_map_get_fd(int ufd)
+{
+	return fdget(ufd);
+}
+
+EXPORT_SYMBOL(accel_map_get_fd);
+
+int accel_map_lookup_elem(struct fd f, void *key, void *value, u64 flags)
+{
+	struct bpf_map *map;
+
+	/*
+	 *		**BPF_F_LOCK**
+ 	 *			Look up the value of a spin-locked map without
+ 	 *			returning the lock. This must be specified if the
+ 	 *			elements contain a spinlock.
+	*/
+
+	/* Don't need this anymore
+	if (CHECK_ATTR(BPF_MAP_LOOKUP_ELEM))
+		return -EINVAL; */
+
+	if (flags & ~BPF_F_LOCK) {
+		printk("accel_map_lookup_elem: Wrong flag given.\n");
+		return -EINVAL;
+	}
+
+	/* We need to resolve the kernel map FD at the start instead of each time
+	f = fdget(ufd); */
+	map = __bpf_map_get(f);
+	if (IS_ERR(map))
+		return PTR_ERR(map);
+	if (!(map_get_sys_perms(map, f) & FMODE_CAN_READ)) {
+		printk("accel_map_lookup_elem: Map permissions not satisfied.\n");
+		return -EPERM;
+	}
+
+	if ((flags & BPF_F_LOCK) &&
+	    !map_value_has_spin_lock(map)) {
+		printk("accel_map_lookup_elem: Map does NOT have spinlock.\n");
+		return -EINVAL;
+	}
+
+	/* Don't need to copy key from userspace anymore
+	key = __bpf_copy_key(ukey, map->key_size);
+	if (IS_ERR(key)) {
+		err = PTR_ERR(key);
+		goto err_put;
+	} */
+
+	/* Size of value is not needed anymore
+	value_size = bpf_map_value_size(map); */
+
+	/* bpf_map_copy_value should read directly into _value
+	err = -ENOMEM;
+	value = kmalloc(value_size, GFP_USER | __GFP_NOWARN);
+	if (!value)
+		goto free_key; */
+
+	return bpf_map_copy_value(map, key, value, flags);
+	/* Don't need gotos anymore
+	if (err)
+		goto free_value; */
+
+	/* Don't need to copy back to userspace
+	err = -EFAULT;
+	if (copy_to_user(uvalue, value, value_size) != 0)
+		goto free_value; */
+
+	/* err = 0; */
+
+/* No need to free any variables, since they are all
+   owned by the caller
+free_value:
+	kfree(value);
+free_key:
+	kfree(key);
+err_put:
+	fdput(f);
+	return err; */
+}
+
+EXPORT_SYMBOL(accel_map_lookup_elem);
+
 
 #define BPF_MAP_UPDATE_ELEM_LAST_FIELD flags
 
